DESIGN-DECISIONS.md

```diff
@@ -0,0 +1,342 @@
+in time.
+
+Jumping across Rust stack frames is unsound at best, and completely unsafe at worst.  Specifically, doing so defeats any
+Rust destructor that was planned to execute during normal stack unwinding.  The ramifications of this can vary from 
+leaking memory to leaving things like reference-counted objects or locks completely unaware that they've been released.
+If Rust's stack does not properly unwind, it is impossible to know the program state after `sigsetjmp`'s second return.
+
+With this in mind, it's also important to know that any Postgres internal function is subject to raise an ERROR, causing
+a `longjmp`.
+
+To solve this, pgx creates its own `sigsetjmp` points at each Postgres FFI boundary. As part of pgx' machine-generated 
+Rust bindings, it wraps each Postgres function, adding its own `sigsetjmp` handling prior to calling the internal 
+Postgres function.  The wrapper lies and tells Postgres that if it raises an error it should `longjmp` to this 
+`sigsetjmp` point instead of the one Postgres created at transaction start.  Regardless of if an ERROR is raised, pgx 
+restores Postgres' understanding of the original `sigsetjmp` point after the internal function returns.  These wrapper 
+functions are what's exposed to the user, ensuring it's not possible to bypass this protection.
+
+(as an aside, I did some quick LLVM IR callgraph analysis a few years ago and found that, indeed, nearly every Postgres
+function can raise an ERROR (either directly or indirectly through its callgraph).  I don't recall the specific numbers,
+but it was such a small percentage that didn't call `ereport()` that I didn't even consider adding such smarts to pgx' 
+bindings generator)
+
+When Postgres `longjmp`s due to an ERROR during a function call made through pgx, its wrapper function's `sigsetjmp` 
+point activates and the ERROR is converted into a Rust `panic!()`.  This panic is then raised using Rust's normal panic 
+machinery.  As such, the Rust stack unwinds properly, ensuring all Rust destructors are called.
+
+After Rust's stack has unwound, the `panic!()` is handed off to `ereport()`, passing final error handling control back 
+to Postgres.
+
+This process is similar to Postgres' `PG_TRY`/`PG_CATCH` C macros.  However, it happens at *every* Rust-->Postgres 
+function call boundary, allowing Rust to unwind its stack and call destructors before relinquishing control to 
+Postgres.
+
+### Protecting Postgres Transactions (read: asking Rust for help)
+
+The above is about how to handle Postgres ERRORs when an internal Postgres function is called from Rust.  This section
+talks about how to handle Rust `panic!()`s generated by a Rust function called from Postgres.
+
+A decision was made during initial pgx development that **all** Rust `panic!()`s will properly abort the active
+Postgres transaction.  While Rust's preference is to abort the **process**, this makes no sense in the context of a
```

Comment:
As above, this isn't quite right.

Reply:
I changed it above, but I'm just gonna rewrite this as "terminate the **process**".  I don't really want to have to explain threads again.  The document having a bit of a conversational tone, I think folks reading it will understand.